include/range/v3/detail/config.hpp

```diff
@@ -22,15 +22,42 @@
 #define RANGES_ASSERT assert
 #endif
 
+#if defined(__GNUC__) || defined(__clang__)
+#define RANGES_ASSUME(COND) ((COND) ? void() : __builtin_unreachable())
+#elif defined(_MSC_VER)
+#define RANGES_ASSUME(COND) __assume(COND)
+#endif
+
+#ifndef RANGES_CONTRACT_VIOLATIONS_THROW
+#define RANGES_CONTRACT_VIOLATIONS_THROW 0
+#endif
+
 #ifndef RANGES_ENSURE_MSG
-#include <exception>
-#define RANGES_ENSURE_MSG(COND, MSG) \
-    ((COND) ? void() : (RANGES_ASSERT(!(true && MSG)), std::terminate()))
+#if RANGES_CONTRACT_VIOLATIONS_THROW
+#include <stdexcept>
+namespace ranges
+{
+    inline namespace v3
+    {
+        struct contract_violation
```

Comment:
> It's not How Things Are Done (tm) in range-v3

I'm very much not a fan of throwing on contract violations. This was a haphazard attempt to preserve more of the existing test structure; it didn't occur to me that the throwing violations backdoor for this test makes it possible for this garbage to leak in to the rest of the library. And yes, I think that choosing to support it here means it will eventually contaminate the rest of the library.

I agree this isn't a road we want to point ourselves down.